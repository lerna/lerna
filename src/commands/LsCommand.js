import Command from "../Command";
import chalk from "chalk";

export default class LsCommand extends Command {
  initialize(callback) {
    // Nothing to do...
    callback(null, true);
  }

  execute(callback) {
    const formattedPackages = this.filteredPackages
      .map((pkg) => `- ${pkg.name}${pkg.isPrivate() ? ` (${chalk.red("private")})` : ""}`)
      .join("\n");

    this.logger.info(formattedPackages);
    callback(null, true);
  }
}
