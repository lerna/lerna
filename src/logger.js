import progressBar from "./progressBar";
import chalk from "chalk";
import pad from "pad";

const cwd = process.cwd();

const DEFAULT_LOGLEVEL = "info";

const LEVELS = [
  [ "silly",   "magenta" ],
  [ "verbose", "blue"    ],
  [ "info",    "reset"   ],
  [ "success", "green"   ],
  [ "warn",    "yellow"  ],
  [ "error",   "red"     ],
  [ "silent",            ],
];

const TYPE_TO_LEVEL = LEVELS
  .reduce((map, [type], index) => (map[type] = index, map), {});

class Logger {
  constructor() {
    this.setLogLevel();
    this.logs = [];
  }

  setLogLevel(type) {
    this.loglevel = TYPE_TO_LEVEL[type || DEFAULT_LOGLEVEL];
  }

  _log(type, style, level, message, error) {
    this.logs.push({
      type,
      message,
      error
    });

    if (level < this.loglevel) {
      return;
    }

    if (error) {
      message += "\n" + (error.stack || error);
    }

    if (style) {
      message = style(message);
    }

    progressBar.clear();
    this._emit(message);
    progressBar.restore();
  }

  _emit(message) {
    if (process.env.NODE_ENV !== "lerna-test") {
      console.log(message);
    }
  }

  newLine() {
    this._emit("");
  }

  logifyAsync() {
    return (target, property, descriptor) => {
      const message = target.name + "." + property;
      const method = descriptor.value;

      descriptor.value = (...args) => {
        const callback = args.pop();
        const msg = this._formatMethod(message, args);

        this.verbose(msg);

        // wrap final callback
        args.push((error, value) => {
          if (error) {
            this.error(msg, error);
          } else {
            this.verbose(msg + " => " + this._formatValue(value));
          }

          callback(error, value);
        });

        method(...args);
      };
    };
  }

  logifySync() {
    return (target, property, descriptor) => {
      const message = target.name + "." + property;
      const method = descriptor.value;

      descriptor.value = (...args) => {
        const msg = this._formatMethod(message, args);

        this.verbose(msg);

        try {
          const result = method(...args);
          this.verbose(msg + " => " + this._formatValue(result));
          return result;
        } catch (error) {
          this.error(msg, error);
          throw error;
        }
      };
    };
  }

  _formatMethod(method, args) {
    return pad(method, 30, " ") + "(" + this._formatArguments(args) + ")";
  }

  _formatArguments(args) {
    const fullArgs = args.map(this._formatValue).join(", ");
    if (fullArgs.length > 100) {
      return fullArgs.slice(0, 100) + "...";
    } else {
      return fullArgs;
    }
  }

  _formatValue(arg) {
    if (typeof arg === "function") {
      return "function " + arg.name + "() {...}";
    }

    return (JSON.stringify(arg) || "").replace(cwd, ".");
  }
}

LEVELS.forEach(([type, color]) => {
  if (!color) return; // "silent"
  const style = chalk[color];
  const level = TYPE_TO_LEVEL[type];
  Logger.prototype[type] = function(message, error) {
    this._log(type, style, level, message, error);
  };
});

export default new Logger();
