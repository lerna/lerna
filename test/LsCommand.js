import assert from "assert";
import chalk from "chalk";

import exitWithCode from "./_exitWithCode";
import initFixture from "./_initFixture";
import LsCommand from "../src/commands/LsCommand";
import logger from "../src/logger";
import stub from "./_stub";

describe("LsCommand", () => {
  beforeEach((done) => {
    initFixture("LsCommand/basic", done);
  });

  it("should list changes", (done) => {
    const lsCommand = new LsCommand([], {});

    lsCommand.runValidations();
    lsCommand.runPreparations();

    stub(logger, "info", (message) => {
      assert.equal(message, `- package-1\n- package-2\n- package-3\n- package-4\n- package-5 (${chalk.red("private")})`);
    });

    lsCommand.runCommand(exitWithCode(0, done));
  });

  // Both of these commands should result in the same outcome
  const filters = [
    { test: "should list changes for a given scope", flag: "scope", flagValue: "package-1"},
    { test: "should not list changes for ignored packages", flag: "ignore", flagValue: "package-@(2|3|4|5)"},
  ];
  filters.forEach((filter) => {
    it(filter.test, (done) => {
      const lsCommand = new LsCommand([], {[filter.flag]: filter.flagValue});

      lsCommand.runValidations();
      lsCommand.runPreparations();

      stub(logger, "info", (message) => {
        assert.equal(message, "- package-1");
      });

      lsCommand.runCommand(exitWithCode(0, done));
    });
  });
});
