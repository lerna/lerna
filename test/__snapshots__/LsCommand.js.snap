// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LsCommand in a basic repo should list changes for a given scope 1`] = `
Array [
  "package-1 v1.0.0 ",
]
`;

exports[`LsCommand in a basic repo should list packages 1`] = `
Array [
  "package-1 v1.0.0          
package-2 v1.0.0          
package-3 v1.0.0          
package-4 v1.0.0          
package-5 v1.0.0 (private)",
]
`;

exports[`LsCommand in a basic repo should not list changes for ignored packages 1`] = `
Array [
  "package-1 v1.0.0 ",
]
`;

exports[`LsCommand in a repo with packages outside of packages/ should list packages 1`] = `
Array [
  "package-1 v1.0.0 
package-2 v1.0.0 
package-3 v1.0.0 ",
]
`;

exports[`LsCommand with --include-filtered-dependencies should list packages, including filtered ones 1`] = `
Array [
  "@test/package-2 v1.0.0 
@test/package-1 v1.0.0 ",
]
`;
